src/telemetry-tracing.ts

```diff
@@ -234,84 +279,123 @@ export class PubsubSpans {
     message?: PubsubMessage
   ): SpanAttributes {
     const destinationName = params.topicName ?? params.subName;
-    if (!destinationName || (params.topicName && params.subName)) {
-      throw new Error('One of topicName or subName must be specified');
-    }
-    const destinationKind = params.topicName ? 'topic' : 'subscription';
+    const destinationId = params.topicId ?? params.subId;
+    const projectId = params.projectId;
+
+    // Purposefully leaving this debug check here as a comment - this should
+    // always be true, but we don't want to fail in prod if it's not.
+    /*if (
+      (params.topicName && params.subName) ||
+      (!destinationName && !projectId && !destinationId)
+    ) {
+      throw new Error(
+        'One of topicName or subName must be specified, and must be fully qualified'
+      );
+    }*/
 
     const spanAttributes = {
       // Add Opentelemetry semantic convention attributes to the span, based on:
       // https://github.com/open-telemetry/opentelemetry-specification/blob/v1.1.0/specification/trace/semantic_conventions/messaging.md
-      [SemanticAttributes.MESSAGING_TEMP_DESTINATION]: false,
-      [SemanticAttributes.MESSAGING_SYSTEM]: 'pubsub',
-      [SemanticAttributes.MESSAGING_OPERATION]: 'send',
-      [SemanticAttributes.MESSAGING_DESTINATION]: destinationName,
-      [SemanticAttributes.MESSAGING_DESTINATION_KIND]: destinationKind,
-      [SemanticAttributes.MESSAGING_PROTOCOL]: 'pubsub',
+      ['messaging.system']: 'gcp_pubsub',
+      ['messaging.destination.name']: destinationId ?? destinationName,
+      ['gcp.project_id']: projectId,
     } as SpanAttributes;
 
     if (message) {
-      if (message.data?.length) {
-        spanAttributes[
-          SemanticAttributes.MESSAGING_MESSAGE_PAYLOAD_SIZE_BYTES
-        ] = message.data?.length;
+      if (message.calculatedSize) {
+        spanAttributes['messaging.message.envelope.size'] =
+          message.calculatedSize;
+      } else {
+        if (message.data?.length) {
+          spanAttributes['messaging.message.envelope.size'] =
+            message.data?.length;
+        }
       }
       if (message.orderingKey) {
-        spanAttributes['messaging.gcp_pubsub.message.ordering_key'] =
+        spanAttributes['messaging.gcp_pubsub.ordering_key'] =
```

Comment:
Actually the original was right: https://opentelemetry.io/docs/specs/semconv/messaging/gcp-pubsub/

Fixed in the doc